File: greedy/섬연결하기.py
def find_parent(parents, x):
    if parents[x] != x:
        parents[x] = find_parent(parents, parents[x])
    return parents[x]

def union(parents, a, b):
    a = find_parent(parents, a)
    b = find_parent(parents, b)
    if a < b:
        parents[b] = a
    else:
        parents[a] = b
    
        
def solution(n, costs):
    parents = [0] * (n+1)
    for i in range(n+1):
        parents[i] = i
    costs.sort(key=lambda x: x[2])
    answer = 0
    for i in costs:
        a, b, cost = i
        if find_parent(parents, a) != find_parent(parents, b):
            union(parents, a, b)
            answer += cost
    
    return answer

File: greedy/섬연결하기.java
import java.util.Arrays;
import java.util.stream.IntStream;
// import java.util.Comparator;

class Solution {
    public int[] parent;
    public int solution(int n, int[][] costs) {
        int answer = 0;
        parent = IntStream.range(0, n).toArray();
        
        Arrays.sort(costs, (o1, o2) -> o1[2] - o2[2]);
        // Arrays.sort(costs, new Comparator<int[]>() {
        //     @Override
        //     public int compare(int[] o1, int[] o2) {
        //         Integer u = o1[2];
        //         Integer v = o2[2];
        //         return u.compareTo(v);
        //     }
        // });
        
        for (int[] cost : costs) {
            int x = cost[0];
            int y = cost[1];
            int c = cost[2];
            x = findParent(x);
            y = findParent(y);
            if (x != y) {
                answer += c;
                unionParent(x, y);
            }
        }
        return answer;
    }
    
    public int findParent(int v) {
        if (parent[v] != v) {
            return parent[v] = findParent(parent[v]);
        }
        return parent[v];
    }
    
    public void unionParent(int v, int u) {
        v = findParent(v);
        u = findParent(u);
        if (v > u) {
            parent[v] = u;
        } else {
            parent[u] = v;
        }
    }
}
